rust/uefi/linux-bootloader/src/measure.rs

```diff
@@ -37,43 +38,93 @@ pub fn measure_image(system_table: &SystemTable<Boot>, image: &PeInMemory) -> ue
-                        section_name,
-                    )? {
-                        measurements += 1;
-                    }
-                }
-            }
+
+    // Match behaviour of systemd-stub (see src/boot/efi/stub.c in systemd)
+    // The encoding as well as the ordering of measurements is critical.
+    //
+    // References:
+    //
+    // "TPM2 PCR Measurements Made by systemd", https://systemd.io/TPM2_PCR_MEASUREMENTS/
+    //   Section: PCR Measurements Made by systemd-stub (UEFI)
+    //   - PCR 11, EV_IPL, “PE Section Name”
+    //   - PCR 11, EV_IPL, “PE Section Data”
+    //
+    // Unified Kernel Image (UKI) specification, UAPI Group,
+    // https://uapi-group.org/specifications/specs/unified_kernel_image/#uki-tpm-pcr-measurements
+    //
+    // Citing from "UKI TPM PCR Measurements":
+    //   On systems with a Trusted Platform Module (TPM) the UEFI boot stub shall measure the sections listed above,
+    //   starting from the .linux section, in the order as listed (which should be considered the canonical order).
+    //   The .pcrsig section is not measured.
+    //
+    //   For each section two measurements shall be made into PCR 11 with the event code EV_IPL:
+    //    - The section name in ASCII (including one trailing NUL byte)
+    //    - The (binary) section contents
+    //
+    //   The above should be repeated for every section defined above, so that the measurements are interleaved:
+    //   section name followed by section data, followed by the next section name and its section data, and so on.
+
+    // NOTE: The order of measurements is important, so the use of BTreeMap is intentional here.
+    let ordered_sections: BTreeMap<_, _> = pe
+        .sections
+        .iter()
+        .filter_map(|section| {
+            let section_name = section.name().ok()?;
+            let unified_section = UnifiedSection::try_from(section_name).ok()?;
+            unified_section
+                .should_be_measured()
+                .then_some((unified_section, section))
+        })
+        .collect();
+
+    for (unified_section, section) in ordered_sections {
+        let section_name = unified_section.name();
+
+        info!("Measuring section `{}`...", section_name);
+
+        // First measure the section name itself
+        // This needs to be an UTF-8 encoded string with a trailing null byte
+        //
+        // As per reference:
+        // "Measured hash covers the PE section name in ASCII (including a trailing NUL byte!)."
+        let section_name_cstr_utf8 = unified_section.name_cstr();
+
+        if tpm_log_event_ascii(
+            boot_services,
+            TPM_PCR_INDEX_KERNEL_IMAGE,
+            section_name_cstr_utf8.as_bytes_with_nul(),
+            section_name,
+        )? {
+            measurements += 1;
+        }
+
+        // Then measure the section contents.
+        let Some(data) = pe_section_data(pe_binary, section) else {
+            continue;
+        };
```

Comment:
Is this a case where we should continue? Or should we just fail?

```suggestion
        let data = pe_section_data(pe_binary, section).with_context(|| "Failed to extract section: {section_name}")?;
```